gadget/validate.go

```diff
@@ -159,98 +182,88 @@ func validateReservedLabels(vs *VolumeStructure) error {
 	// TODO: should we validate usage of uc20 specific system-recovery-{image,select}
 	//       roles too? they should only be used on uc20 systems, so models that
 	//       have a grade set and are not classic
 
 	switch {
-	case state.SystemSeed == nil && state.SystemData == nil:
-		if wantsSystemSeed(model) {
+	case roles[SystemSeed] == nil && roles[SystemData] == nil:
+		if expectedSeed {
 			return fmt.Errorf("model requires system-seed partition, but no system-seed or system-data partition found")
 		}
-	case state.SystemSeed != nil && state.SystemData == nil:
+	case roles[SystemSeed] != nil && roles[SystemData] == nil:
 		return fmt.Errorf("the system-seed role requires system-data to be defined")
-	case state.SystemSeed == nil && state.SystemData != nil:
+	case roles[SystemSeed] == nil && roles[SystemData] != nil:
 		// error if we have the SystemSeed constraint but no actual system-seed structure
-		if wantsSystemSeed(model) {
+		if expectedSeed {
 			return fmt.Errorf("model requires system-seed structure, but none was found")
 		}
 		// without SystemSeed, system-data label must be implicit or writable
-		if err := checkImplicitLabel(SystemData, state.SystemData, implicitSystemDataLabel); err != nil {
+		if err := checkImplicitLabel(SystemData, roles[SystemData].s, implicitSystemDataLabel); err != nil {
 			return err
 		}
-	case state.SystemSeed != nil && state.SystemData != nil:
+	case roles[SystemSeed] != nil && roles[SystemData] != nil:
 		// error if we don't have the SystemSeed constraint but we have a system-seed structure
-		if !wantsSystemSeed(model) {
+		if !expectedSeed {
 			return fmt.Errorf("model does not support the system-seed role")
 		}
-		if err := checkSeedDataImplicitLabels(state); err != nil {
+		if err := checkSeedDataImplicitLabels(roles); err != nil {
 			return err
 		}
 	}
-	if state.SystemSave != nil {
-		if err := ensureSystemSaveRuleConsistency(state); err != nil {
+	if roles[SystemSave] != nil {
+		if !expectedSeed {
+			return fmt.Errorf("model does not support the system-save role")
+		}
+		if err := ensureSystemSaveRuleConsistency(roles); err != nil {
 			return err
 		}
 	}
-	return nil
-}
 
-func checkImplicitLabel(role string, vs *VolumeStructure, implicitLabel string) error {
-	if vs.Label != "" && vs.Label != implicitLabel {
-		return fmt.Errorf("%s structure must have an implicit label or %q, not %q", role, implicitLabel, vs.Label)
+	if expectedSeed {
+		// make sure that all roles come from the same volume
+		// TODO:UC20: there is more to do in order to support multi-volume situations
+
+		// if SystemSeed is unset we must have failed earlier
+		seedVolName := roles[SystemSeed].volName
 
+		for _, otherRole := range []string{SystemBoot, SystemData, SystemSave} {
+			ri := roles[otherRole]
+			if ri != nil && ri.volName != seedVolName {
+				return fmt.Errorf("system-boot|data|save are expected to share the same volume as system-seed")
```

Comment:
minor preference to have a longer error message here, but not a big deal either way

```suggestion
				return fmt.Errorf("system-boot, system-data, and system-|save are expected to share the same volume as system-seed")
```